### PR TITLE
Different text input length limits for different fields

### PR DESCRIPTION
**Pull request by BartoszKedziorek**
This is the PR that I declared to make. https://github.com/OpenLoco/OpenLoco/pull/1646#discussion_r976794802

This PR basicly makes text input to not accept more characters  than maximal length of the string that it's setting. I'm not sure about the limit for "enter_host_address", so i assigned it with 15 (length of the longest possible IPv4 addess). "chat_title" and "title_industry_name" are mapped with 15 by now, like every other "title". I haven't checked limit for "chat_title" yet. Text input window don't open when I double-click the caption in the industry window (idk if it's a bug).

### CODE REVIEW

File: src/OpenLoco/Windows/TextInputWindow.cpp [hunk 1]
```diff
@@ -41,6 +42,20 @@ namespace OpenLoco::Ui::Windows::TextInput
...
+        { StringIds::scenario_name_title, 63 },
+        { StringIds::scenario_details_title, 199 },
+        { StringIds::title_station_name, 31 },
+        { StringIds::chat_title, 31 },
+        { StringIds::title_industry_name, 31 },
+        { StringIds::enter_host_address, 15 },
+        { StringIds::title_town_name, 31 },
+        { StringIds::title_name_vehicle, 31 }
+    };
+
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Windows/TextInputWindow.cpp:58]: This feels a little hacky. I think it would be better to have the invoking window pass the limit, rather than mapping limits based on a string id.

Comment (LeftofZen) [reply src/OpenLoco/Windows/TextInputWindow.cpp:58]: Agreed

Comment (duncanspumpkin) [reply src/OpenLoco/Windows/TextInputWindow.cpp:58]: And all the 31s should use the kMaxUserStringLength-1 or whatever we named the constant

Comment (AaronVanGeffen) [reply src/OpenLoco/Windows/TextInputWindow.cpp:58]: > And all the 31s should use the kMaxUserStringLength-1 or whatever we named the constant

Perhaps we can even make this the default, as it seems to be the most common.

File: src/OpenLoco/Localisation/StringManager.h [hunk 2]
```diff
@@ -135,6 +136,11 @@ namespace OpenLoco
 
 namespace OpenLoco::StringManager
 {
+    const uint8_t kUserStringSize = 32;
+
+    const uint16_t kUserStringsStart = 0x8000;
+    const uint16_t kUserStringsEnd = kUserStringsStart + Limits::kMaxUserStrings;
+
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Localisation/StringManager.h:143]: It's fine to expose `kUserStringSize`, but please revert the change to `kUserStringsStart` and `kUserStringsEnd`.

File: enc_temp_folder/39f0c3ddf3f0a4fe42f94aa6a54a8516/TextInput.cpp [hunk 3]
```diff
@@ -0,0 +1,189 @@
+#include "../Ui/TextInput.h"
```
Comment (AaronVanGeffen) [inline enc_temp_folder/39f0c3ddf3f0a4fe42f94aa6a54a8516/TextInput.cpp:1]: Please remove this temp file.

File: src/OpenLoco/Ui/TextInput.cpp [hunk 4]
```diff
@@ -166,4 +166,25 @@ namespace OpenLoco::Ui::TextInput
...
+
+            if (callerTitle == StringIds::scenario_details_title)
+            {
+                return uint16_t(std::size(mOptions.scenarioDetails) - 1);    
+            }
+
+            if (callerTitle == StringIds::scenario_name_title)
+            {
+                return uint16_t(std::size(mOptions.scenarioName) - 1);
+            }
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Ui/TextInput.cpp:184]: I really think this should be up to the caller to _set_ (e.g. a window function that initiates a text input session). The InputSession shouldn't have to _infer_ it based on a string.

Comment (BartoszKedziorek) [reply src/OpenLoco/Ui/TextInput.cpp:184]: Would it be bad if the caller determine length by stringsIds or should I find different solution?

Comment (AaronVanGeffen) [reply src/OpenLoco/Ui/TextInput.cpp:184]: You can assume the caller has access to this information. Consider, for example:
https://github.com/OpenLoco/OpenLoco/blob/71813b28e73f54451a6ce9311cccdde0ff186167/src/OpenLoco/Windows/CompanyWindow.cpp#L2626-L2630

You could change this to include another parameter for input size:
```cpp
        static void renameCompanyPrompt(Window* self, WidgetIndex_t widgetIndex)
        {
            auto company = CompanyManager::get(CompanyId(self->number));
            auto inputSize = StringManager::kUserStringSize;
            TextInput::openTextInput(self, StringIds::title_name_company,
                StringIds::prompt_enter_new_company_name, company->name,
                //  /------------ new parameter
                // V
                inputSize, widgetIndex, nullptr);
        }
```

File: src/OpenLoco/Ui/TextInput.h [hunk 5]
```diff
@@ -13,14 +14,16 @@ namespace OpenLoco::Ui::TextInput
         size_t cursorPosition; // 0x01136FA2
         int16_t xOffset;       // 0x01136FA4
         uint8_t cursorFrame;   // 0x011370A9
+        int inputLenLimit;
 
         InputSession() = default;
-        InputSession(const std::string initialString)
+        InputSession(const std::string initialString, int inputSize = StringManager::kUserStringSize)
```
Comment (duncanspumpkin) [inline src/OpenLoco/Ui/TextInput.h:20]: Make it a uint32_t and the default should be kUserStringSize-1.

File: src/OpenLoco/Ui/WindowManager.h [hunk 6]
```diff
@@ -343,7 +343,7 @@ namespace OpenLoco::Ui::Windows
     {
         void registerHooks();
 
-        void openTextInput(Ui::Window* w, string_id title, string_id message, string_id value, int callingWidget, void* valueArgs);
+        void openTextInput(Ui::Window* w, string_id title, string_id message, string_id value, int inputSize, int callingWidget, void* valueArgs);
```
Comment (duncanspumpkin) [inline src/OpenLoco/Ui/WindowManager.h:346]: uint32_t

File: src/OpenLoco/Windows/CompanyWindow.cpp [hunk 7]
```diff
@@ -297,7 +297,8 @@ namespace OpenLoco::Ui::Windows::CompanyWindow
                 case widx::change_owner_name:
                 {
                     auto company = CompanyManager::get(CompanyId(self.number));
-                    TextInput::openTextInput(&self, StringIds::title_name_owner, StringIds::prompt_enter_new_name_for_owner, company->ownerName, widgetIndex, nullptr);
+                    auto inputSize = StringManager::kUserStringSize - 1;
+                    TextInput::openTextInput(&self, StringIds::title_name_owner, StringIds::prompt_enter_new_name_for_owner, company->ownerName, inputSize, widgetIndex, nullptr);
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Windows/CompanyWindow.cpp:301]: One final comment (hopefully). We have made `StringManager::kUserStringSize - 1` the default for `InputSession`. I think would make sense to instead make it the default for `TextInput::openTextInput`. This also means moving the parameter to the last position, but that should be an easy operation in most IDEs.

This `inputSize` parameter can then be omitted everywhere where `StringManager::kUserStringSize - 1` is intended to be the limit. In other words, only in three places will you actually need to pass the parameter.

Sorry, I don't mean to

File: src/OpenLoco/Ui/TextInput.h [hunk 8]
```diff
@@ -1,4 +1,5 @@
 #include "../Graphics/Gfx.h"
+#include "../Localisation/StringManager.h"
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Ui/TextInput.h:2]: This include is no longer needed.

**Review by AaronVanGeffen — APPROVED** [review]
